OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,274 @@
+    virtual size_t  getNumCols() const = 0;
+
+    /** Determine if the DataTable has a column associated with a 
+        column label (string) */
+    virtual bool hasColumn(const std::string& label) const = 0;
+    /** For a known column label return its column index. */
+    virtual size_t getColumnIndex(const std::string& label) const = 0;
+
+    /** Read-only column labels associated with the columns of the DataTable */
+    virtual const Array<std::string>& getColumnLabels() const = 0;
+    /** Update column labels associated with the columns of the DataTable */
+    virtual Array<std::string>& updColumnLabels() = 0;
+
+    /* TODO: Read in the units associated with the columns of the DataTable */
+    /* virtual const Array<Units>& getColumnUnits() const = 0;
+    */
+
+    /** Get the underlying type of data held in the DataTable. Can be any 
+        SimTK numerical type: Real, Vec<M>, Mat<M, N>, SpatialVec, ... */
+    virtual const std::type_index& getDataTypeInfo() const = 0;
+
+    /** Dump the DataTable to an output stream */
+    virtual void dumpToStream(std::ostream &out) const = 0;
+
+//=============================================================================
+};  // END of class AbstractDataTable
+//=============================================================================
+
+
+/** Concrete Implementation of the <tt>DataTable<DataType></tt>
+@tparam DataType  Any numerical type: (SimTK::) Real, Vec<M>, Mat<M,N>, ...*/
+template<typename DataType = SimTK::Real>
+class DataTable_ : public AbstractDataTable {
+//=============================================================================
+// METHODS
+//=============================================================================
+public:
+    virtual ~DataTable_(){};
+
+    /** Construct an empty DataTable */
+    DataTable_() : _type(typeid(DataType)) {}
+
+    /** Construct a pre-sized DataTable to be filled in later. If values
+        are accessed before they are assigned they will be NaN. */
+    DataTable_(size_t nrows, size_t ncols, 
+        const DataType& val = DataType(SimTK::NaN)) :
+        _data(int(nrows), int(ncols), val),
+        _type(typeid(DataType)) {}
+
+    /** Convenience construction of a DataTable from a filename with its source
+        type identified by the file extension. E.g. `results.csv`, `states.sto`
+    @param[in] filename         the name of the data file to be accessed
+    */
+    DataTable_(const std::string& filename) :
+        _type(typeid(DataType)) {
+        std::string daId = FileAdapter::findExtension(filename);
+        auto* reader = FileAdapter::createAdapter(daId);
+        if (!reader){
+            throw Exception("DataTable() no adapter to read file '"
+                + filename + "' could be found.");
+        }
+        reader->setFilename(filename);
+        reader->prepareForReading(*this);
+        reader->read();
+        delete reader;
+    }
+
+    DataTable_& operator=(const DataTable_& dt) {
```

Comment:
Getting rid of the `_type` method will allow you to use the default copy assignment and delete this implementation.